lib/devicetrust/native/device_darwin.m

```diff
@@ -0,0 +1,249 @@
+// go:build darwin
+
+#include "device_darwin.h"
+
+#import <CoreFoundation/CoreFoundation.h>
+#import <Foundation/Foundation.h>
+#import <IOKit/IOKitLib.h>
+#import <Security/Security.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+const char *kDeviceKeyLabel = "com.gravitational.teleport.devicekey";
+
+const int kErrNoApplicationTag = 1;
+const int kErrNoValueRef = 2;
+const int kErrCopyPubKeyFailed = 3;
+
+OSStatus findDeviceKey(NSNumber *retAttrs, CFTypeRef *out) {
+  NSData *label = [NSData dataWithBytes:(void *)kDeviceKeyLabel
+                                 length:strlen(kDeviceKeyLabel)];
+  NSDictionary *query = @{
+    (id)kSecClass : (id)kSecClassKey,
+    (id)kSecAttrKeyType : (id)kSecAttrKeyTypeECSECPrimeRandom,
+    (id)kSecMatchLimit : (id)kSecMatchLimitOne,
+    (id)kSecReturnRef : @YES,
+    (id)kSecReturnAttributes : (id)retAttrs,
+    (id)kSecAttrApplicationLabel : label,
+  };
+  return SecItemCopyMatching((CFDictionaryRef)query, out);
+}
+
+void copyPublicKey(const char *id, CFDataRef pubKeyRep, PublicKey *pubKeyOut) {
+  pubKeyOut->id = strdup(id);
+  pubKeyOut->pub_key_len = CFDataGetLength(pubKeyRep);
+  pubKeyOut->pub_key = calloc(pubKeyOut->pub_key_len, sizeof(uint8_t));
+  memcpy(pubKeyOut->pub_key, CFDataGetBytePtr(pubKeyRep),
+         pubKeyOut->pub_key_len);
+}
+
+int64_t DeviceKeyCreate(const char *newID, PublicKey *pubKeyOut) {
+  CFErrorRef err = NULL;
+  SecAccessControlRef access = NULL;
+  NSString *label = NULL;     // managed by ARC
+  NSData *labelAsData = NULL; // managed by ARC
+  NSData *idAsData = NULL;    // managed by ARC
+  NSDictionary *attrs = NULL; // managed by ARC
+  SecKeyRef privKey = NULL;
+  SecKeyRef pubKey = NULL;
+  CFDataRef pubKeyRep = NULL;
+
+  int64_t res = DeviceKeyGet(pubKeyOut);
+  if (res != errSecItemNotFound) {
```

Comment:
Some comments about the control flow - here and elsewhere, when checking errors - would be welcome.

Reply:
Added a comment.